### ISSUE TITLE
Activity Page does not support Filtering and Pagination together correctly

### ISSUE THREAD

**Issue by PeachesMLG** [jellyfin/jellyfin-web#6221]
**Describe The Bug**
The Activity Page does not support Filtering and Pagination together due to doing pagination on server side, and filtering on the client side.
This means If you set a filter for a user, it only filters the results on the current page you are on.
e.g. Page1 may return 0 results, Page 2 may return 2 results, page 3 5 ect.

**Steps To Reproduce**
This is demenstrated best on an account that doesnt use jellyfin often
1. Go to the activity page
2. Create a filter (e.g. filter by a user)
3. Navigate through the pages and see the problem

**Screenshots**
Page 1
![image](https://github.com/user-attachments/assets/efe18edf-b035-4bc3-8e45-35bdb25ac7c0)
Page 2
![Untitled design](https://github.com/user-attachments/assets/2fe9de4f-f4c0-4010-b885-4173323a9559)

**System (please complete the following information):**
 - Platform: Browser
 - Browser: Firefox
 - Jellyfin Version: 10.9.11

**Comment by A7Lavinraj** [comment]
That means filtering should be done first before pagination, right?

**Comment by thornbill** [comment]
It's definitely not ideal, but this is currently working as intended. Pagination happens on the backend because the query can be slow (I have 126,000+ entries on my main server) and filtering happens on the frontend because the api does not support filtering. We might be able to add some logic to fetch additional rows if filtering results in less than the page size but that doesn't appear to be something that is supported "out of the box" with the table library.

**Comment by PeachesMLG** [comment]
> It's definitely not ideal, but this is currently working as intended. Pagination happens on the backend because the query can be slow (I have 126,000+ entries on my main server) and filtering happens on the frontend because the api does not support filtering. We might be able to add some logic to fetch additional rows if filtering results in less than the page size but that doesn't appear to be something that is supported "out of the box" with the table library.

When you say the API doesnt support it, do you mean the Jellyfin server API? or the Material React Table Library?

I think the best option here would be to add filtering onto the backend Jellyfin api no?

**Comment by thornbill** [comment]
The user activity API in the Jellyfin server does not support filtering.

**Comment by A7Lavinraj** [comment]
So what we have to build anyways, like give some hints to do it

**Comment by PeachesMLG** [comment]
> The user activity API in the Jellyfin server does not support filtering.

So from what I can see it has Date and UserId filtering
https://github.com/jellyfin/jellyfin/blob/master/Jellyfin.Api/Controllers/ActivityLogController.cs#L49
I guess it potentially needs a Level and Type filter aswell? Is there anything else the backend would need in order for this to work that im missing?

**Comment by Dan-613** [comment]
Just thinking about working on this and adding filtering server-side, however it might be slow to return results if the Activity Logs are big since the Type is stored as a string unless we add an index on the type.

Is it worth it to add filtering server-side if an index is required?

**Comment by thornbill** [comment]
> So from what I can see it has Date and UserId filtering
> 
> https://github.com/jellyfin/jellyfin/blob/master/Jellyfin.Api/Controllers/ActivityLogController.cs#L49
> 
> I guess it potentially needs a Level and Type filter aswell? Is there anything else the backend would need in order for this to work that im missing?

This is not correct. You can only currently filter on system vs user events and get the activity since a specific date.

**Comment by Dan-613** [comment]
> This is not correct. You can only currently filter on system vs user events and get the activity since a specific date.

The suggestion is to add Type and Level filtering on the server-side.

**Comment by viown** [comment]
> Just thinking about working on this and adding filtering server-side, however it might be slow to return results if the Activity Logs are big since the Type is stored as a string unless we add an index on the type.

It might be a bit too complex to implement the required functionality to make it work on web as we'd need support for both sorting and filtering (for all columns) by query.

**Comment by Dan-613** [comment]
> It might be a bit too complex to implement the required functionality to make it work on web as we'd need support for both sorting and filtering (for all columns) by query.

I don't mind giving it a try. Material React Table being used seems to support server-side querying and filtering.

**Comment by Speiger** [comment]
Ever considered making the filtering maybe client sided until the server can do it?
I mean you need admin privileges anyways so its not even something that is like "don't trust the client" type of issue.

**Comment by claude-cp** [comment]
@crobibero I see you recently had a PR in regards to this issue? 
https://github.com/jellyfin/jellyfin/pull/15583

Does this effectively solve the issue?
Which part? As per my understanding, some changes were needed both on the server, as well as on the client?
Will it be released with the next JF server version?

Sorry for my lack of understanding the JF code and its current state, but quite curious what's going on with this issue.

**Comment by crobibero** [comment]
unstable version of the server (10.12+) now supports sorting and filtering server-side
https://github.com/jellyfin/jellyfin/pull/15583

**Comment by claude-cp** [comment]
Yes and, any changes needed to the client side too, or?

**Comment by crobibero** [comment]
> [@crobibero](https://github.com/crobibero) I see you recently had a PR in regards to this issue? [jellyfin/jellyfin#15583](https://github.com/jellyfin/jellyfin/pull/15583)
> 
> Does this effectively solve the issue? Which part? As per my understanding, some changes were needed both on the server, as well as on the client? Will it be released with the next JF server version?
> 
> Sorry for my lack of understanding the JF code and its current state, but quite curious what's going on with this issue.

Haha you commented while I was writing comment. Timing is everything.

I implemented the server-side part of this feature, now the web client can change the activity log table to use server-side paging and filtering. I looked at implementing this myself but I've never used react.

I believe the only place changes would need to be made is here: https://github.com/jellyfin/jellyfin-web/blob/master/src/apps/dashboard/routes/activity/index.tsx

**Comment by bjorntp** [comment]
Great, this has been bothering me for months and has been meaning to set up the API endpoints, just when I was finally getting to it you had done it two weeks earlier!

I have a mostly working frontend now, got some minor kinks to and cleanup to fix, the MRT-table was pretty easy to change to a backend-filtering instead. Since the backend changes are still only in the unstable branch, can I still create a PR for the frontend changes?

If so I'll do within the coming week.

**Comment by bjorntp** [comment]
@crobibero 

I'm not sure if this is the correct place to discuss the API but since the PR is closed I figured we might as well do it here. I do not not think it will be a long discussion.

~I think there might be something wrong the multi-sort functionality in the backend. While the API states that it can take several when CSV are sent in only one of them is applied. Seems like it is always a certain prioritization as well.~
~For example these requests will give the exact same response where the sorting is based on ShortOverview and Name is ignored.~
~`...&sortBy=Name,ShortOverview&sortOrder=Ascending,Descending`~
~`...&sortBy=ShortOverview,Name&sortOrder=Descending,Ascending`~
~This is me checking the responses, not shows how they are shown in the UI.~
~As I understand the API this would would be the correct formatting but feel free to correct me if I am wrong, I am not familiar enough with C# to be certain while reading the code either.~

I did start reading some of the source and realized the that the formatting I used in the beginning was correct, it has issues as well though.
`...sortBy=ShortOverview&sortBy=Name&sortOrder=Descending&sortOrder=Descending`
It only applies the last one in the Url, so in this it would be sorted based on `Name`.

Filters seem to work as intended. Although maxDate endpoint would be nice for range-filters. Both can be implemented

**Comment by bjorntp** [comment]
It would be great if someone could test the changes made in #7431. From my perspective everything works well, but since I introduced some minor UI changes, I’d really appreciate a second opinion. It would be helpful if someone with a larger database could try it out as well.

**Comment by Mumbolio85** [comment]
I thought this would be a good issue to start looking at as I want to contribute to jellyfin this summer. Since all that was needed to do on this issue was test this issue on a large dataset. In order to test this, I modified useLogEntries.ts to generate a specifiable amount of log entries (i started with 500), that are all different. 

Upon initial testing, it apeared to work when sorting by level (info, warning, error, ect.) and so I expanded the dataset to 10,000 to do more of a stress test. It continued to work on every page that I checked and I tested it with multiple different filters applied.

If using mock data to test this did not create other problems (of which I saw none), I believe that this issue can be closed.

Here is a screenshot of the activity page with a filter applied on a large dataset, I checked multiple other pages and they all had 25 rows per page.

<img width="1198" height="890" alt="Image" src="https://github.com/user-attachments/assets/ce5ce640-da4c-4632-85a7-821138f542a2" />

**Comment by pmarks-net** [comment]
@Mumbolio85 which version did you test?  I still see the bug on 10.11.10, but it sounds like updating to 10.12+ might help.

**Comment by Mumbolio85** [comment]
I'm running 10.11.10 on both the jellyfin server (running on macos in a docker container to allow for an in progress jellyfin-web to be run on top of it) as well as the front end jellyfin-web (not including the changes needed to generate the mock data).

**Comment by pmarks-net** [comment]
@Mumbolio85 your screenshot shows filtering by Level, but I can't say if that works here because they're all "Information".  I notice the bug when filtering by User or Name, which may be different because they're strings.

**Comment by thornbill** [comment]
This was a new feature added for the upcoming 12.0 release.
